[Sim Jun Heng] iP

simjunheng opened the pull request:
# Duke2k22
> “Life is too complicated not to be orderly.” 😀 – Martha Stewart [(source)](https://dansilvestre.com/productivity-quotes/)

## Why should you use Duke2k22
- text-based
- easy to learn
- ~~FAST~~ *ULTRA FAST* to use

## How to get started
1. `download` from [here](https://github.com/simjunheng/ip/releases/tag/A-Jar)
2. `double-click` it 
3. `add` your tasks
4. sit back and relax 😘 

Not to mention, it's **FREE**

## Key Features
```ruby
- [x] Manage Tasks
- [ ] Add Reminders(coming soon)
```

Commit:
Add Gradle support

simjunheng committed:
Level 1. Greet, Echo, Exit

simjunheng committed:
Level 2. Add, List

simjunheng committed:
Level 3. Mark as Done

simjunheng committed:
Revert "Level 3. Mark as Done"

This reverts commit 5cae4e12e1a5abfc8357f9fffa7cb2e604019ec7.

simjunheng committed:
Revert "Revert "Level 3. Mark as Done""

This reverts commit 317bb4bf29957eb9a1a4aa5b388c2e15935b60a5.

simjunheng committed:
Level 3. Mark as Done

simjunheng committed:
Level 3. Mark as Done

simjunheng committed:
Level 4. ToDos, Events, Deadlines

simjunheng committed:
A-TextUiTesting: Automated Text UI Testing

simjunheng committed:
Level 5. Handle Errors

simjunheng committed:
Level 5. Handle Errors

simjunheng committed:
Level 6. Delete

simjunheng committed:
Change greeting message for duke program

simjunheng committed:
Add help command for new users

simjunheng committed:
Edit addToDo, addDeadline, addEvent after removing rmType and extractDateORTask

simjunheng committed:
Change getTask to toString

simjunheng committed:
Edit Bug in addToDo, addDeadline, addEvent method

simjunheng committed:
Add new save feature (level 7)

simjunheng committed:
Add DateTime feature

simjunheng committed:
Merge branch 'save-feature'

simjunheng committed:
Merge branch 'dt-feature'

simjunheng committed:
Add parser, storage, tasklist, ui class

simjunheng committed:
Put all classes into duke package

simjunheng committed:
Create junit test for parser and task class

simjunheng committed:
Add javadocs to class methods

simjunheng committed:
Tweak code to comply with coding standard

simjunheng committed:
Add find feature

simjunheng committed:
Merge branch 'branch-A-JavaDoc'

simjunheng committed:
Merge branch 'branch-A-CodingStandard'

simjunheng committed:
Merge branch 'branch-Level-9'

simjunheng committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

simjunheng committed:
Change file path in gradle

simjunheng committed:
Refactor file after checkstyle

simjunheng committed:
Merge branch 'A-CheckStyle'

kxshxsh commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @param desc the task description.
+     * @param date the date of the event.
+     * @param isDone the current completion status of the task.
```
Really appreciate how your javadocs follow conventions to a T!

kxshxsh commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,59 @@
+        String str = "[E]";
+        // Convert Current Date Format to MMM dd yyyy
+        String customPattern = "MMM dd yyyy";
+        DateTimeFormatter customFormat = DateTimeFormatter.ofPattern(customPattern);
```
Just a suggestion, but in my opinion, I think you can skip line 31 and directly alter line 32 to be `DateTimeFormatter customFormat = DateTimeFormatter.ofPattern("MMM dd yyyy");`

I don't think it would break the logical flow or make it less comprehensive. Maybe this is something you can consider if you agree :)

kxshxsh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,58 @@
+            int index = 1;
+            while (index < temp.length) {
+                if (!temp[index].equals("/by") && !temp[index].equals("/at")) {
+                    if (desc == null) {
```
An alternative approach could be to use `str.indexOf(...)` to search for "/by" and "/at" and then use `str.substring(...)` to help extract the different sections of the input string. I think it could help save on the while loops and also help improve the readability of your code. What do you think?

kxshxsh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,197 @@
+    /**
+     * Adds to do task into our current list.
+     *
+     * @param desc the task description entered by the user
```
Something small that you have been doing consistently otherwise but have seemed to miss out here is the period after the parameter description in your javadocs. Very tiny detail but something good to be conscious about:)

kxshxsh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,197 @@
+     */
+    public void findTask(String taskName) {
+        int num = 1;
+        int matchNum = 0;
```
Subjectively, I think `matchNum` can be named something clearer, more descriptive and detailed:) Such as `numOfMatches`

kxshxsh commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @throws DukeException if invalid inputs are detected.
+     */
+    public static void notSpecified() throws DukeException {
```
Our coding standard says that "Names representing methods must be verbs and written in camelCase." Perhaps you can consider changing the name of `notSpecified()` to fit the conventions better?

kxshxsh reviewed (commented):
Hi!! I think you have done a great job in your project:) Everything seems well thought out and elaborate! Your javadocs are also near flawless! There are, however, some instances where I think there is an easier alternative so I have pointed them out for you to consider and evaluate!

ryanwalterlee commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+        DateTimeFormatter customFormat = DateTimeFormatter.ofPattern(customPattern);
+        String tempDate = customFormat.format(this.date);
+        if (super.isDone) {
+            str += "[X] " + super.desc + " (by: " + tempDate + ")";
+        } else {
+            str += "[ ] " + super.desc + " (by: " + tempDate + ")";
+        }
```
In my opinion, I prefer to write strings in java using template literals, such as 
```
String.format("[X] %s by: %s)", super.desc, tempDate);
```
This way, it is not as easy to miss out a " " (space), which could lead to bugs that could take hours to resolve. If this syntax interests you, you could try this next time :)

s7manth replied:
I agree with this. 

In fact, I created a custom `OutputFormatter` just for a bulk `append` method for `StringBuilder`.

ryanwalterlee commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Assigns desc, date and done to this instance.
+     *
+     * @param desc the task description
+     * @param date the date of the event
+     * @param isDone the current completion status of the task.
+     */
```
According to the coding standard stated here [link](https://se-education.org/guides/conventions/java/intermediate.html), it states that all javadoc should have the following form where each @params line ends with a fullstop. IsDone has a fullstop at the end so perhaps you overlooked the previous 2.

simjunheng committed:
Add GUI to duke program

simjunheng committed:
Merge branch 'branch-Level-10'

ryanwalterlee commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,197 @@
+            throw new DukeException("PLEASE INCLUDE THE DATE IN YOUR COMMAND.");
+        }
+        try {
+            LocalDate date2 = LocalDate.parse(date);
```
I think the variable name 'date2' is not very descriptive of what it is. At first look, I don't exactly know what date2 is. After inspecting the code, I realised that date2 is actually a parsed date. Perhaps the variable name could be called dateParsed instead? I understand variable naming is difficult when there are more important logic to handle but every small effort goes a long way for the readability of code.

simjunheng committed:
Remove import from DialogBox

ryanwalterlee reviewed (commented):
Overall good coding standard observed, had a hard time trying to look for violations.

From what I see throughout, some things to take note of:
1. javadoc comments ends with '.'
2. variable naming could be more descriptive (try thinking in a strangers shoes to determine if someone else can understand)

s7manth commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,53 @@
+        list = Storage.loadFile();
+    }
+
+    public String getResponse(String input) throws IOException {
```
Instead of using so many if-else statements, you could have abstracted out the handling commands part. `Parser` identifies the type of command, and maybe `Ui` could have handled the command.

s7manth commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+     * @return an arraylist containing data read form the file.
+     * @throws IOException if there's error arising from the file methods
+     * @throws DukeException if data in the file is corrupted (not in the correct storage format)
+     */
```
Great Java Docs!

s7manth commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+     * @throws IOException if there's error arising from I/O methods.
+     */
+    public static void writeFile(TaskList list) throws IOException {
+        FileWriter fw = new FileWriter("./data/duke.txt");
```
Should create the directory and the file, if they didn't exist to avoid `FileNotFoundException`.

s7manth reviewed (commented):
Overall, the code was of good quality. LGTM

simjunheng committed:
Improve code quality and OOP

Improve readability for reviewers as long methods, insufficient SLAP, and arrowhead style code is observed.

Added FileParser and InputParser class (subclass of Parser) to reduce the line of codes in the getResponse() method in Duke and loadFile() method in Storage. Renamed method names
to make it more intuitive for the code reviewers. Shifted methods (mainly the commands) from TaskList to Ui.

Unable to make Parser class universal without inheritance, which is why FileParser and InputParser subclass are created. Methods (mainly the commands) should not be placed in TaskList, but instead should be placed in Ui since this commands deals with input from the users and returns an output string to the interface.

simjunheng committed:
Add assert in Storage class

simjunheng committed:
Merge pull request #1 from simjunheng/branch-A-Assertions

Add assert in Storage class

simjunheng committed:
Merge branch 'master' into branch-A-CodeQuality

simjunheng committed:
Merge pull request #2 from simjunheng/branch-A-CodeQuality

Improve code quality and OOP

simjunheng committed:
Add missing resources and resolve conflicts

simjunheng committed:
Add, attach and save tags to task

simjunheng committed:
Fix error in TaskTest test class and improve code quality for InputParser class

simjunheng committed:
Rename commands to make it more intuitive

simjunheng committed:
Change background image

simjunheng committed:
Give proper credits to se.edu for javafx classes and fix minor bugs in gui fxml file.

simjunheng committed:
Add javafx application header

simjunheng committed:
Add app representative image into docs folder

simjunheng committed:
Edit readme.md
